Lib/dataclasses.py

```diff
@@ -1267,6 +1268,32 @@ def is_dataclass(obj):
     return hasattr(cls, _FIELDS)
 
 
+class DataclassLike(metaclass=abc.ABCMeta):
+    """Abstract base class for all dataclass types.
+
+    Mainly useful for type-checking.
+    """
+    # __dataclass_fields__ here is really an "abstract class variable",
+    # but there's no good way of expressing that at runtime,
+    # so just make it a regular class variable with a dummy value
+    __dataclass_fields__ = {}
+
+    def __init_subclass__(cls):
+        raise TypeError(
+            "Use the @dataclass decorator to create dataclasses, "
+            "rather than subclassing dataclasses.DataclassLike"
+        )
+
+    def __new__(cls):
+        raise TypeError(
+            "dataclasses.DataclassLike is an abstract class that cannot be instantiated"
+        )
+
+    @classmethod
+    def __subclasshook__(cls, other):
+        return hasattr(other, _FIELDS)
```

Comment:
I would prefer this code:

```suggestion
        return is_dataclass(other)
```

It is slightly slower but has more semantics.

Reply:
I'm -0 on this, as I don't think it reads significantly better, so I don't see the reason for making it slower. But don't mind changing it if others agree.

Reply:
In principle I think it's nicer to not duplicate the definition of `is_dataclass`. In practice I think that definition is unlikely to ever change, so it doesn't matter much. I am curious how much difference it actually makes to the performance of `isinstance(SomeDataclass, DataclassLike)`; if it's not really significant I'd probably go for `is_dataclass`, if it is I'd take the speed. I don't think it's a make-or-break issue for this PR either way :)

Reply:
How about refactoring `is_dataclass()`:

```python
def _is_dataclass_type(cls):
    """Returns True if cls is a dataclass."""
    return hasattr(cls, _FIELDS)


def _is_dataclass_instance(obj):
    """Returns True if obj is an instance of a dataclass."""
    return _is_dataclass_type(type(obj))


def is_dataclass(obj):
    """Returns True if obj is a dataclass or an instance of a
    dataclass."""
    cls = obj if isinstance(obj, type) else type(obj)
    return _is_dataclass_type(cls)
```

Then you could use:
```python
    @classmethod
    def __subclasshook__(cls, other):
        return _is_dataclass_type(other)
```

Reply:
> How about refactoring `is_dataclass()`:

Hmm, that feels out of scope for this PR -- I'm uncomfortable refactoring other parts of `dataclasses.py` for a nice-to-have-but-inessential feature, which is basically what this is :)

Reply:
> I am curious how much difference it actually makes to the performance of `isinstance(SomeDataclass, DataclassLike)`; if it's not really significant I'd probably go for `is_dataclass`, if it is I'd take the speed. I don't think it's a make-or-break issue for this PR either way :)

Okay, I benchmarked it, and the difference between the two is in the noise :) So I'll accept @sobolevn's suggestion here!